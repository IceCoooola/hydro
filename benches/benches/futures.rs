use std::cell::RefCell;
use std::future::Future;
use std::pin::Pin;
use std::rc::Rc;
use std::task::{Context, Poll, Waker};
use std::time::Duration;

use criterion::{criterion_group, criterion_main, Criterion};
use hydroflow::hydroflow_syntax;
use hydroflow::scheduled::graph::Hydroflow;

const NUM_ELEMS: u32 = 3000;

/// A future which returns () after it manually woken
pub struct ManualFut {
    done: Rc<RefCell<bool>>,
    waker: Rc<RefCell<Option<Waker>>>,
}
impl ManualFut {
    pub fn new(done: Rc<RefCell<bool>>, waker: Rc<RefCell<Option<Waker>>>) -> ManualFut {
        ManualFut { done, waker }
    }
}

impl Future for ManualFut {
    type Output = ();

    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        match *self.done.borrow() {
            true => Poll::Ready(()),
            false => {
                self.waker.replace(Some(cx.waker().clone()));
                Poll::Pending
            }
        }
    }
}

fn benchmark_immediately_available(c: &mut Criterion) {
    c.bench_function("futures/immediately_available", |b| {
        b.iter_batched(
            || {
                let mut df = hydroflow_syntax! {
                    source_iter(0..NUM_ELEMS)
                    -> map(|x| async move {
                        x
                    })
                    -> defer_tick()
                    -> poll_futures()
                    -> for_each(|_| {});
                };

                df.run_tick(); // skip loading and mapping to future
                df
            },
            |mut df| {
                df.run_tick();
            },
            criterion::BatchSize::SmallInput,
        );
    });
}

type WakeInfo = (Rc<RefCell<bool>>, Vec<Rc<RefCell<Option<Waker>>>>);

fn benchmark_delayed(c: &mut Criterion) {
    fn setup<'a>(count: u32, wake_one: bool) -> (Hydroflow<'a>, WakeInfo) {
        let done = Rc::new(RefCell::new(false));
        let mut wakers = Vec::new();

        let range = 0..count;
        let futs = range
            .map(|i| {
                let waker = Rc::new(RefCell::new(None));
                let d = if !wake_one || i == 0 {
                    wakers.push(waker.clone());
                    done.clone()
                } else {
                    Rc::new(RefCell::new(false))
                };
                ManualFut::new(d, waker)
            })
            .collect::<Vec<_>>();

        let df = {
            hydroflow_syntax! {
                source_iter(futs)
                -> poll_futures()
                -> for_each(|_| {});
            }
        };

        (df, (done, wakers))
    }

    fn wake_all((done, wakers): WakeInfo) {
        *done.borrow_mut() = true;
        wakers.into_iter().for_each(|waker| {
            if let Some(waker) = waker.borrow_mut().take() {
                waker.wake();
            } else {
                panic!("waker not found but future should have been polled")
            }
        })
    }

    // Tick with the initial poll
    c.bench_function("futures/delayed/initial", |b| {
        b.iter_batched(
            || setup(NUM_ELEMS, false).0,
            |mut df| {
                df.run_tick();
            },
            criterion::BatchSize::SmallInput,
        );
    });

    // Tick when no results are available
    c.bench_function("futures/delayed/waiting", |b| {
        b.iter_batched(
            || {
                let (mut df, wakes) = setup(NUM_ELEMS, true);
                df.run_tick();
                df.run_tick();
                df.run_tick();
                wake_all(wakes);
                df
            },
            |mut df| {
                df.run_tick();
            },
            criterion::BatchSize::SmallInput,
        );
    });

    // Tick when results became available
    c.bench_function("futures/delayed/ready", |b| {
        b.iter_batched(
            || {
                let (mut df, wakes) = setup(NUM_ELEMS, false);
                df.run_tick();
                wake_all(wakes);
                df
            },
            |mut df| {
                df.run_tick();
            },
            criterion::BatchSize::SmallInput,
        );
    });
    // Tick after all results have been consumed
    c.bench_function("futures/delayed/done", |b| {
        b.iter_batched(
            || {
                let (mut df, wakes) = setup(NUM_ELEMS, false);
                df.run_tick();
                wake_all(wakes);
                df.run_tick();
                df
            },
            |mut df| {
                df.run_tick();
            },
            criterion::BatchSize::SmallInput,
        );
    });
}

criterion_group!(
    name=futures;
    config=Criterion::default().measurement_time(Duration::from_secs(30));
    targets=benchmark_immediately_available,
    benchmark_delayed
);
criterion_main!(futures);
