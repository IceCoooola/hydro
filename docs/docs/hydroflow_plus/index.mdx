---
sidebar_position: 0
---

# Introduction
Hydroflow+ layers a high-level Rust API over the Hydroflow IR, making it possible to write dataflow programs that span multiple processes with straightline, functional Rust code. Hydroflow+ is built on top of [Stageleft](./stageleft.mdx), which allows Hydroflow+ to emit regular Hydroflow programs that are compiled into efficient Rust binaries. It also integrates with [Hydro Deploy](../deploy/index.md) to make it easy to deploy and run Hydroflow+ programs on a cluster.

The main logic of Hydroflow+ programs manipulates **streams**, which capture infinite ordered sequences of elements. Streams are transformed using classic functional operators such as `map`, `filter`, and `fold`, as well as relational operators such as `join`. To build **distributed** dataflow programs, Hydroflow+ also introduces the concept of **processes**, which capture _where_ a stream is being processed.

## Setup
Hydroflow+ requires a particular workspace setup, as any crate that uses Hydroflow+ must have an supporting macro crate to drive the code generation. To get started, we recommend using the Hydroflow+ template.

```bash
#shell-command-next-line
cargo install cargo-generate
#shell-command-next-line
cargo generate hydro-project/hydroflow-plus-template
```

Then, you can test the example dataflow:

```bash
#shell-command-next-line
cargo test
```
