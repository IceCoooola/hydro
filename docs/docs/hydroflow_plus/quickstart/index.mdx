# Quickstart
In this tutorial, we'll walk through the basics of Hydroflow+ by building a simple dataflow that prints out the first 10 natural numbers. We'll start with a single process, then pipeline the computation, and finally distribute it across a cluster.

To get started with a new project, we'll use the Hydroflow+ template. The template comes with a pre-configured build system and an implementation of the following examples.

```bash
#shell-command-next-line
cargo install cargo-generate
#shell-command-next-line
cargo generate hydro-project/hydroflow-plus-template
# Project Name: my-example-project
#shell-command-next-line
cd my-example-project
```
