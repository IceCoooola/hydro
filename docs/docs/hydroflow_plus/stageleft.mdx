---
title: Stageleft
sidebar_position: 6
---

import StageleftDocs from '../../../stageleft/README.md'

<StageleftDocs/>
