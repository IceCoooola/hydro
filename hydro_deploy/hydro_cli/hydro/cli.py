import hydro._core # type: ignore
import sys

def cli():
    try:
        hydro._core.cli.cli_entrypoint(sys.argv)
        return 0
    except Exception:
        return 1
