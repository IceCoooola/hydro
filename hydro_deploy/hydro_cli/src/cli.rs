use std::fmt::Display;
use std::path::PathBuf;

use anyhow::Context;
use clap::{Parser, Subcommand};
use pyo3::exceptions::PyException;
use pyo3::prelude::*;
use pyo3::types::PyList;

use crate::{AnyhowError, AnyhowWrapper};

#[derive(Parser, Debug)]
#[command(name = "Hydro Deploy", author, version, about, long_about = None)]
struct Cli {
    #[command(subcommand)]
    command: Commands,
}

#[derive(Subcommand, Debug)]
enum Commands {
    /// Deploys an application given a Python deployment script.
    Deploy {
        /// Path to the deployment script.
        config: PathBuf,
        /// Additional arguments to pass to the deployment script.
        #[arg(last(true))]
        args: Vec<String>,
    },
}

fn async_wrapper_module(py: Python) -> Result<&PyModule, PyErr> {
    PyModule::from_code(
        py,
        include_str!("../hydro/async_wrapper.py"),
        "wrapper.py",
        "wrapper",
    )
}

#[derive(Debug)]
struct PyErrWithTraceback {
    err_display: String,
    traceback: String,
}

impl std::error::Error for PyErrWithTraceback {}

impl Display for PyErrWithTraceback {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f, "{}", self.err_display)?;
        write!(f, "{}", self.traceback)
    }
}

fn deploy(config: PathBuf, args: Vec<String>) -> anyhow::Result<()> {
    Python::with_gil(|py| -> anyhow::Result<()> {
        let syspath: &PyList = py
            .import("sys")
            .and_then(|s| s.getattr("path"))
            .and_then(|p| Ok(p.downcast::<PyList>()?))?;

        syspath.insert(0, PathBuf::from(".").canonicalize().unwrap())?;

        let filename = config.canonicalize().unwrap();
        let fun: Py<PyAny> = PyModule::from_code(
            py,
            std::fs::read_to_string(config).unwrap().as_str(),
            filename.to_str().unwrap(),
            "",
        )
        .with_context(|| format!("failed to load deployment script: {}", filename.display()))?
        .getattr("main")
        .context("expected deployment script to define a `main` function, but one was not found")?
        .into();

        let wrapper = async_wrapper_module(py)?;
        match wrapper.call_method1("run", (fun, args)) {
            Ok(_) => Ok(()),
            Err(err) => {
                let traceback = err
                    .traceback(py)
                    .context("traceback was expected but none found")
                    .and_then(|tb| Ok(tb.format()?))?
                    .trim()
                    .to_string();

                if err.is_instance_of::<AnyhowError>(py) {
                    let args = err
                        .value(py)
                        .getattr("args")?
                        .extract::<Vec<AnyhowWrapper>>()?;
                    let wrapper = args.first().unwrap();
                    let underlying = &wrapper.underlying;
                    let mut underlying = underlying.blocking_write();
                    Err(underlying.take().unwrap()).context(traceback)
                } else {
                    Err(PyErrWithTraceback {
                        err_display: format!("{}", err),
                        traceback,
                    }
                    .into())
                }
            }
        }
    })?;

    Ok(())
}

#[pyfunction]
fn cli_entrypoint(args: Vec<String>) -> PyResult<()> {
    match Cli::try_parse_from(args) {
        Ok(args) => {
            let res = match args.command {
                Commands::Deploy { config, args } => deploy(config, args),
            };

            match res {
                Ok(_) => Ok(()),
                Err(err) => {
                    eprintln!("{:?}", err);
                    Err(PyErr::new::<PyException, _>(""))
                }
            }
        }
        Err(err) => {
            err.print().unwrap();
            Err(PyErr::new::<PyException, _>(""))
        }
    }
}

#[pymodule]
pub fn cli(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(cli_entrypoint, m)?)?;

    Ok(())
}
