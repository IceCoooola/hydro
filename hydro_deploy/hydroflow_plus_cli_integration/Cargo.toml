[package]
name = "hydroflow_plus_cli_integration"
publish = true
version = "0.7.0"
edition = "2021"
license = "Apache-2.0"
documentation = "https://docs.rs/hydroflow_plus_cli_integration/"
description = "Library for working with hydro_deploy and hydroflow_plus"

[features]
default = []
deploy = [ "hydro_deploy", "async-channel" ]

[dependencies]
stageleft = { path = "../../stageleft", version = "^0.3.0" }
hydroflow_plus = { path = "../../hydroflow_plus", version = "^0.7.0", features = [ "cli_integration" ] }
syn = { version = "2.0.0", features = [ "parsing", "extra-traits" ] }
tokio = { version = "1.16", features = [ "full" ] }
serde = { version = "1", features = [ "derive" ] }

hydro_deploy = { path = "../core", version = "^0.7.0", optional = true }
async-channel = { version = "1.8.0", optional = true }

# added to workaround `cargo smart-release` https://github.com/Byron/cargo-smart-release/issues/16
stageleft_tool = { path = "../../stageleft_tool", version = "^0.2.0", optional = true }

[build-dependencies]
stageleft_tool = { path = "../../stageleft_tool", version = "^0.2.0" }
