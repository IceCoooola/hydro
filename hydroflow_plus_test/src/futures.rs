use std::time::Duration;

use hydroflow_plus::tokio::sync::mpsc::UnboundedSender;
use hydroflow_plus::*;
use stageleft::{q, Quoted, RuntimeData};

#[stageleft::entry]
pub fn unordered<'a>(
    flow: FlowBuilder<'a, SingleProcessGraph>,
    output: RuntimeData<&'a UnboundedSender<u32>>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let process = flow.process(&());

    flow.source_iter(&process, q!([2, 3, 1, 9, 6, 5, 4, 7, 8]))
        .map(q!(|x| async move {
            // tokio::time::sleep works, import then just sleep does not, unsure why
            tokio::time::sleep(Duration::from_millis(10)).await;
            x
        }))
        .poll_futures()
        .for_each(q!(|x| output.send(x).unwrap()));

    flow.extract().optimize_default()
}

#[stageleft::entry]
pub fn ordered<'a>(
    flow: FlowBuilder<'a, SingleProcessGraph>,
    output: RuntimeData<&'a UnboundedSender<u32>>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let process = flow.process(&());

    flow.source_iter(&process, q!([2, 3, 1, 9, 6, 5, 4, 7, 8]))
        .map(q!(|x| async move {
            // tokio::time::sleep works, import then just sleep does not, unsure why
            tokio::time::sleep(Duration::from_millis(10)).await;
            x
        }))
        .poll_futures_ordered()
        .for_each(q!(|x| output.send(x).unwrap()));

    flow.extract().optimize_default()
}

#[stageleft::runtime]
#[cfg(test)]
mod tests {
    use std::collections::HashSet;
    use std::time::Duration;

    use hydroflow_plus::util::collect_ready_async;

    #[tokio::test]
    async fn test_unordered() {
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut flow = super::unordered!(&out);
        let handle = tokio::task::spawn(async move {
            tokio::time::sleep(Duration::from_secs(1)).await;
            assert_eq!(
                HashSet::from_iter(1..10),
                collect_ready_async::<HashSet<_>, _>(&mut out_recv).await
            );
        });

        tokio::time::timeout(Duration::from_secs(2), flow.run_async())
            .await
            .expect_err("Expected time out");

        handle.await.unwrap();
    }

    #[tokio::test]
    async fn test_ordered() {
        let (out, mut out_recv) = hydroflow_plus::util::unbounded_channel();

        let mut flow = super::ordered!(&out);
        let handle = tokio::task::spawn(async move {
            tokio::time::sleep(Duration::from_secs(1)).await;
            assert_eq!(
                Vec::from_iter([2, 3, 1, 9, 6, 5, 4, 7, 8]),
                collect_ready_async::<Vec<_>, _>(&mut out_recv).await
            );
        });

        tokio::time::timeout(Duration::from_secs(2), flow.run_async())
            .await
            .expect_err("Expected time out");

        handle.await.unwrap();
    }
}
