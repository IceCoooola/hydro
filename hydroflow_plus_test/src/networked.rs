use hydroflow_plus::bytes::Bytes;
use hydroflow_plus::util::cli::HydroCLI;
use hydroflow_plus::*;
use hydroflow_plus_cli_integration::{CLIRuntime, HydroflowPlusMeta};
use stageleft::{q, Quoted, RuntimeData};

#[allow(dead_code)]
pub struct NetworkedBasicIo<'a, D: Deploy<'a>> {
    pub source_zero_port: D::ProcessPort,
    pub process_zero: D::Process,
    pub process_one: D::Process,
    pub cluster_port: D::ClusterPort,
    pub cluster: D::Cluster,
}

pub fn networked_basic<'a, D: Deploy<'a>>(
    flow: &FlowBuilder<'a, D>,
    process_spec: &impl ProcessSpec<'a, D>,
    cluster_spec: &impl ClusterSpec<'a, D>,
) -> NetworkedBasicIo<'a, D> {
    let process_zero = flow.process(process_spec);
    let process_one = flow.process(process_spec);

    let (source_zero_port, source_zero) = flow.source_external(&process_zero);

    source_zero
        .send_bytes(&process_one)
        .for_each(q!(|v: Bytes| {
            println!("node one received: {:?}", std::str::from_utf8(&v).unwrap());
        }));

    let cluster = flow.cluster(cluster_spec);
    let (cluster_port, cluster_stream) = flow.many_source_external::<D::Process, _, _>(&cluster);
    cluster_stream.for_each(q!(|v: Bytes| {
        println!("cluster received: {:?}", std::str::from_utf8(&v).unwrap());
    }));

    NetworkedBasicIo {
        source_zero_port,
        process_zero,
        process_one,
        cluster_port,
        cluster,
    }
}

#[stageleft::entry]
pub fn networked_basic_runtime<'a>(
    flow: FlowBuilder<'a, CLIRuntime>,
    cli: RuntimeData<&'a HydroCLI<HydroflowPlusMeta>>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let _ = networked_basic(&flow, &cli, &cli);
    flow.extract()
        .optimize_default()
        .with_dynamic_id(q!(cli.meta.subgraph_id))
}

#[stageleft::runtime]
#[cfg(test)]
mod tests {
    use std::cell::RefCell;

    use hydro_deploy::{Deployment, HydroflowCrate};
    use hydroflow_plus::futures::SinkExt;
    use hydroflow_plus::util::cli::ConnectedSink;
    use hydroflow_plus_cli_integration::{
        DeployClusterSpec, DeployCrateWrapper, DeployProcessSpec,
    };

    #[tokio::test]
    async fn networked_basic() {
        let mut deployment = Deployment::new();
        let localhost = deployment.Localhost();

        let builder = hydroflow_plus::FlowBuilder::new();
        let deployment = RefCell::new(deployment);
        let io = super::networked_basic(
            &builder,
            &DeployProcessSpec::new(|| {
                deployment.borrow_mut().add_service(
                    HydroflowCrate::new(".", localhost.clone())
                        .bin("networked_basic")
                        .profile("dev"),
                )
            }),
            &DeployClusterSpec::new(|| {
                vec![deployment.borrow_mut().add_service(
                    HydroflowCrate::new(".", localhost.clone())
                        .bin("networked_basic")
                        .profile("dev"),
                )]
            }),
        );

        insta::assert_debug_snapshot!(builder.extract().ir());

        let mut deployment = deployment.into_inner();

        let port_to_zero = io
            .source_zero_port
            .create_sender(&mut deployment, &localhost)
            .await;

        let ports_to_cluster = io
            .cluster_port
            .create_senders(&mut deployment, &localhost)
            .await;

        deployment.deploy().await.unwrap();

        let mut conn_to_zero = port_to_zero.connect().await.into_sink();
        let node_one_stdout = io.process_one.stdout().await;

        let mut conn_to_cluster = ports_to_cluster[0].connect().await.into_sink();
        let cluster_stdout = io.cluster.members[0].stdout().await;

        deployment.start().await.unwrap();

        conn_to_zero.send("hello world!".into()).await.unwrap();
        conn_to_cluster.send("hello cluster!".into()).await.unwrap();

        assert_eq!(
            node_one_stdout.recv().await.unwrap(),
            "node one received: \"hello world!\""
        );

        assert_eq!(
            cluster_stdout.recv().await.unwrap(),
            "cluster received: \"hello cluster!\""
        );
    }
}
