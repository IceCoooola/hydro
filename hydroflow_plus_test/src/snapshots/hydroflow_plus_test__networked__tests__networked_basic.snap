---
source: hydroflow_plus_test/src/networked.rs
expression: builder.build().ir()
---
[
    ForEach {
        f: { use crate :: __staged :: networked :: * ; | v : Bytes | { println ! ("node one received: {:?}" , std :: str :: from_utf8 (& v) . unwrap ()) ; } },
        input: Network {
            to_location: 1,
            serialize_pipeline: None,
            sink_expr: null,
            source_expr: null,
            deserialize_pipeline: Some(
                Operator(
                    Operator {
                        path: "map",
                        args: [
                            "| b | b . unwrap () . freeze ()",
                        ],
                    },
                ),
            ),
            input: Map {
                f: | b | b . unwrap () . freeze (),
                input: Source {
                    source: Stream(
                        null,
                    ),
                    location_id: 0,
                },
            },
        },
    },
    ForEach {
        f: { use crate :: __staged :: networked :: * ; | v : Bytes | { println ! ("cluster received: {:?}" , std :: str :: from_utf8 (& v) . unwrap ()) ; } },
        input: Map {
            f: | b | b . unwrap () . freeze (),
            input: Source {
                source: Stream(
                    null,
                ),
                location_id: 2,
            },
        },
    },
]
